runtime/src/seekable_buffering_reader.rs

```diff
@@ -0,0 +1,195 @@
+use {
+    log::*,
+    solana_measure::measure::Measure,
+    std::{
+        io::*,
+        sync::{
+            atomic::{AtomicBool, AtomicUsize, Ordering},
+            Arc, Mutex, RwLock,
+        },
+        thread::{Builder, JoinHandle},
+    },
+};
+
+pub struct SeekableBufferingReaderInner {
+    // unpacking callers read from 'data'. Data is transferred when 'data' is exhausted.
+    // This minimizes lock contention since bg file reader has to have almost constant write access.
+    pub data: RwLock<Vec<Vec<u8>>>,
+    // bg thread reads to 'new_data'
+    pub new_data: RwLock<Vec<Vec<u8>>>,
+    pub len: AtomicUsize,
+    pub calls: AtomicUsize,
+    pub error: RwLock<std::io::Result<usize>>,
+    pub bg_reader: Mutex<Option<JoinHandle<()>>>,
+    pub stop: AtomicBool,
+}
+
+pub struct SeekableBufferingReader {
+    pub instance: Arc<SeekableBufferingReaderInner>,
+    pub pos: usize,
+    pub last_buffer_index: usize,
+    pub next_index_within_last_buffer: usize,
+}
+
+impl Clone for SeekableBufferingReader {
+    fn clone(&self) -> Self {
+        Self {
+            instance: Arc::clone(&self.instance),
+            pos: 0,
+            last_buffer_index: 0,
+            next_index_within_last_buffer: 0,
+        }
+    }
+}
+
+impl Drop for SeekableBufferingReaderInner {
+    fn drop(&mut self) {
+        if let Some(handle) = self.bg_reader.lock().unwrap().take() {
+            self.stop.store(true, Ordering::Relaxed);
+            handle.join().unwrap();
+        }
+    }
+}
+
+impl SeekableBufferingReader {
+    pub fn new<T: 'static + Read + std::marker::Send>(reader: T) -> Self {
+        let inner = SeekableBufferingReaderInner {
+            new_data: RwLock::new(vec![]),
+            data: RwLock::new(vec![]),
+            len: AtomicUsize::new(0),
+            calls: AtomicUsize::new(0),
+            error: RwLock::new(Ok(0)),
+            bg_reader: Mutex::new(None),
+            stop: AtomicBool::new(false),
+        };
+        let result = Self {
+            instance: Arc::new(inner),
+            pos: 0,
+            last_buffer_index: 0,
+            next_index_within_last_buffer: 0,
+        };
+
+        let result_ = result.clone();
+
+        let handle = Builder::new()
+            .name("solana-compressed_file_reader".to_string())
+            .spawn(move || {
+                result_.read_entire_file_in_bg(reader);
+            });
+        *result.instance.bg_reader.lock().unwrap() = Some(handle.unwrap()); // TODO - unwrap here - do we expect to fail creating a thread? If we do, probably a fatal error anyway.
```

Comment:
I think unwrap is fine, usually its a resource issue. no ram or no mmaps for the stack. Not much we can do to handle it.